### ISSUE TITLE
CoqIde anomaly `Failure("hd")` in Separation Logic Foundations

### ISSUE THREAD

**Issue by rbrott** [coq/coq#14400]
#### Description of the problem

I get `Anomaly "Uncaught exception Failure("hd")."` when stepping through `Basic.v` of [Separation Logic Foundations](https://softwarefoundations.cis.upenn.edu/slf-current/index.html) in CoqIde. Specifically, it occurs during the execution of the `xwp` tactic in the first proof. I cannot reproduce the issue with `coqc`/`coqtop`.

#### Coq Version

```
The Coq Proof Assistant, version 8.13.0 (January 2021)
compiled on Jan 29 2021 0:01:18 with OCaml 4.11.1
```

**Comment by ppedrot** [comment]
Can you print here the terminal output of CoqIDE with the `-debug` flag passed?

**Comment by rbrott** [comment]
```
Anomaly "Uncaught exception Failure("hd")."
Please report at http://coq.inria.fr/bugs/.
Raised at Stdlib.failwith in file "stdlib.ml", line 29, characters 17-33
Called from Stdlib__list.hd in file "list.ml" (inlined), line 30, characters 10-23
Called from Ppconstr.print_hunks.pop in file "printing/ppconstr.ml", line 86, characters 24-34
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 97, characters 16-23
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 121, characters 33-57
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 125, characters 18-23
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 122, characters 18-23
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 121, characters 33-57
Called from Ppconstr.pr_notation in file "printing/ppconstr.ml", line 133, characters 4-52
Called from Ppconstr.pr in file "printing/ppconstr.ml", line 683, characters 8-99
Called from Ppconstr.pr in file "printing/ppconstr.ml", line 681, characters 16-41
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 99, characters 18-75
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 117, characters 18-23
Called from Ppconstr.pr_notation in file "printing/ppconstr.ml", line 133, characters 4-52
Called from Ppconstr.pr in file "printing/ppconstr.ml", line 683, characters 8-99
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 99, characters 18-75
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 117, characters 18-23
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 117, characters 18-23
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 117, characters 18-23
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 98, characters 18-23
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 117, characters 18-23
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 117, characters 18-23
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 121, characters 33-57
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 121, characters 33-57
Called from Ppconstr.pr_notation in file "printing/ppconstr.ml", line 133, characters 4-52
Called from Ppconstr.pr in file "printing/ppconstr.ml", line 683, characters 8-99
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 99, characters 18-75
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 125, characters 18-23
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 117, characters 18-23
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 121, characters 33-57
Called from Ppconstr.pr_notation in file "printing/ppconstr.ml", line 133, characters 4-52
Called from Ppconstr.pr in file "printing/ppconstr.ml", line 683, characters 8-99
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 99, characters 18-75
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 117, characters 18-23
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 117, characters 18-23
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 125, characters 18-23
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 98, characters 18-23
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 117, characters 18-23
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 117, characters 18-23
Called from Ppconstr.print_hunks.aux in file "printing/ppconstr.ml", line 121, characters 33-57
Called from Ppconstr.pr_notation in file "printing/ppconstr.ml", line 133, characters 4-52
Called from Ppconstr.pr in file "printing/ppconstr.ml", line 683, characters 8-99
Called from Idetop.process_goal in file "ide/coqide/idetop.ml", line 200, characters 4-75
Called from CList.map in file "clib/cList.ml", line 413, characters 21-24
Called from Idetop.export_pre_goals in file "ide/coqide/idetop.ml" (inlined), line 224, characters 16-40
Called from Idetop.export_pre_goals in file "ide/coqide/idetop.ml", line 225, characters 31-44
Called from Idetop.goals in file "ide/coqide/idetop.ml", line 241, characters 11-60
Called from Idetop.eval_call.interruptible in file "ide/coqide/idetop.ml", line 456, characters 12-15
Called from Xmlprotocol.abstract_eval_call in file "ide/coqide/protocol/xmlprotocol.ml", line 683, characters 29-46
```

**Comment by mheiber** [comment]
same issue, it seems to be a problem with display of the complex notations in Separation Logic Foundations.


Workarounds:
- fly blind, the proof state still evolves, you just can't see it about half the time. This actually works for the `Basics` chapter since most of the proofs are really similar.
- OR, when you see `Failure("hd")`, do `Unset Printing Notations.` You should be able to see the context and goal again, they will just be hard to read.

**Comment by ejconlon** [comment]
FWIW I ended up downgrading to 8.12, and everything works fine there.

**Comment by ppedrot** [comment]
From the backtrace, I don't think this is CoqIDE-specific, it looks more like the default set of flags of CoqIDE triggers the issue but it should be reproducible with coqc with the right set of options.

**Comment by ppedrot** [comment]
Here is a self-contained test:
```
Set Implicit Arguments.

Axiom var : Type.
Axiom val : Type.
Inductive trm : Type := trm_let : var -> trm -> trm -> trm.
Axiom formula : Type.
Axiom wpgen_let : forall (F1:formula) (F2of:val->formula), formula.

Fixpoint wpgen (t:trm) : formula :=
  match t with
  | trm_let x t1 t2 => wpgen_let (wpgen t1) (fun v => wpgen t2)
  end.

Declare Scope wp_scope.

Declare Custom Entry wp.

Notation "<[ e ]>" :=
  e
  (at level 0, e custom wp at level 99) : wp_scope.

Notation "x" :=
  x
  (in custom wp at level 0,
   x constr at level 0) : wp_scope.

Notation "'Let' x ':=' F1 'in' F2" :=
  ((wpgen_let F1 (fun x => F2)))
  (in custom wp at level 69,
   x ident,
   F1 custom wp at level 99,
   F2 custom wp at level 99,
   right associativity,
  format "'[v' '[' 'Let'  x  ':='  F1  'in' ']' '/' '[' F2 ']' ']'") : wp_scope.

Declare Scope val_scope.

Open Scope wp_scope.

Print wpgen.
```
I realized writing it that this is a duplicate of #14211, which is in particular fixed in 8.14.
